src/ray/core_worker/core_worker.cc

```diff
@@ -459,15 +459,44 @@ CoreWorker::CoreWorker(const CoreWorkerOptions &options, const WorkerID &worker_
       {{"worker_id", worker_id.Hex()}});
 }
 
+CoreWorker::~CoreWorker() { RAY_LOG(INFO) << "Core worker is destructed"; }
+
 void CoreWorker::Shutdown() {
-  io_service_.stop();
+  RAY_LOG(INFO) << "Shutting down a core worker.";
+  is_shutdown_ = true;
   if (options_.worker_type == WorkerType::WORKER) {
+    // Running in a main thread.
     direct_task_receiver_->Stop();
     task_execution_service_.stop();
   }
   if (options_.on_worker_shutdown) {
+    // Running in a main thread.
     options_.on_worker_shutdown(GetWorkerID());
   }
+
+  io_service_.stop();
+  if (gcs_client_) {
+    // TODO(sang): This causes a thread-safety bug because
+    // it invalidates a pointer that's used by a different thread.
```

Comment:
looks previously this logic is called on a different thread; would this cause thread safety issue?

Reply:
That thread safety issue was the root cause of event stats failure. I will fix this in the next PR.